### ISSUE TITLE
Implementation gaps between FlexVolume and CSI for node-local volumes

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#64984]
TLDR: CSI is positioned as our replacement for FlexVolumes.  This issue covers the gaps between existing FlexVolume and existing CSI.  I consider covering these gaps blockers for CSI GA and removing FlexVolume alpha, but fortunately I think they are only small deltas from the existing support.  However, I was unable to implement a CSI plugin that perform node local operations at this time.

FlexVolumes allow extension of nodes for both remote storage (mount, attach, etc) and local storage (exposing filesystem contents from local daemons).  The difference is really the difference between a volume type of "secretRef" and a volume type of "ebsVolume" - the former does not need to be attached to a node, and is implemented by daemon logic solely on the node.  The latter *requires* coordination with an external API, as well as attach and detach logic.

Local volume use cases that FlexVolume can solve that CSI or FlexVolume replacement needs to make possible:

1. Allow a pod to be given a per node identity from a non-Kube identity provider - for example, a local daemon that provides Kerberos or Vault identity tokens as a directory in the container.   

    This is a very opinionated custom secret volume that gets its source from someplace other than Kubernetes.  The volume is explicitly associated with the node, the pod, and the service account.  A cluster level controller might inject a flex volume of type "local.kerberos.identity" into the pod, and the local daemon would then provide a `krb5.keytab` file inside of a tmpfs mounted directory.

2. Allow the contents of an image to be mounted into a pod (https://github.com/dims/docker-flexvol) - so the user defines a volume of type "container.image" that has the image name as a volume parameter ("quay.io/myapplication/static-assets:latest") that is mounted into a directory in my pod (at "/srv/static") so that I don't have to copy them out of an init container.
3. Allow a local unix domain socket to be injected into a pod that provides a node-local API that cannot be intercepted - for instance, someone wants to provide a socket that proxies the Docker daemon but prevents unsafe calls (https://github.com/Tecnativa/docker-socket-proxy).  They would request a volume of type "docker-socket" and a cluster scoped webhook admission plugin would vet the parameters passed to the volume type.
4. Implement a unique local storage solution as an extension, rather than being built in to Kube.

In prototyping such a plugin for 1 (which was a scenario we've been discussing with sig-storage in the container-identity-wg), I ran into the following issues that complicate / prevent these scenarios:

Important:

1. The CSI plugin on the node gets no information about the pod from the Kubelet (which it implicitly trusts).  FlexVolume passes this info:

    ```
	extraOptions[optionKeyPodName] = f.podName
	extraOptions[optionKeyPodNamespace] = f.podNamespace
	extraOptions[optionKeyPodUID] = string(f.podUID)
	// service account metadata
	extraOptions[optionKeyServiceAccountName] = f.podServiceAccountName

	// Extract secret and pass it as options.
	if err := addSecretsToOptions(extraOptions, f.spec, f.podNamespace, f.driverName, f.plugin.host); err != nil {
		os.Remove(dir)
		return err
	}

	// Implicit parameters
	if fsGroup != nil {
		extraOptions[optionFSGroup] = strconv.FormatInt(int64(*fsGroup), 10)
	}
    ```

    Without this info it is not possible for a CSI daemon to, for instance, create a kerberos or other plugin identity specifically for a pod.  The only data the daemon gets is the mount path (which does have the pod UID) but that is a very unsafe assumption and fragile.
    
    I would expect sufficient pod info to be passed to CSI when a plugin is running on Kube that a local implementor can take action based on the pod, rather than on the CSI spec.  This is info that is local to the invocation of the pod (unlike the contents of a CSI spec).

    The simplest suggestion would be to pass additional info (either parts of the pod, or the entire pod) as part of the CSI calls for a given pod on the node, as a platform specific value.  It does not necessarily need to be generic to all CSI implementors, since a Mesos specific CSI plugin would not have the same metadata as Kubernetes.

Annoying / Problematic / Scale issues:

1. Implementing a node local CSI volume requires creating a PV (and running an attacher).  

    For a node local volume, there is no data that lives off the node.  If an implementer expected *every* pod in the system to have a kerberos identity, and there were 100k pods (or 10k namespaces) on the cluster, that would mean that there were 10k/100k PV and 10k/100k PVC for each node local volume.  It also means the implementor must create a controller that creates a fake PV and PVC for every pod, which is both expensive and introduces the possibility that those PVs/PVCs get out of sync.

    Suggest we allow a subset of CSI plugins to be specified in the pod spec (gated by appropriate security controls).  The original reason to not do that was to minimize security implications.  However, now that PodSecurityPolicy can gate FlexVolume use, the same reasoning can be applied to CSI volumes.  Think of the difference between "secretRef" and "AWS" - we're talking about being able to extend things like "secretRef".

2. The concept of "attaching" a node local CSI volume is questionable - node local storage is always attached.

    Assuming we come up with a real solution to the first annoyance above, attachment may not be necessary.

3. It is still difficult to write a clean and correct CSI plugin - more sample code needs to be provided with a reasonable local implementation.  Specifically, mounting volumes requires reimplementing code that could be shared with Kube as a library.

TBD: Add more details from trying this out.

@kubernetes/sig-storage-api-reviews

**Comment by saad-ali** [comment]
Supporting ephemeral volumes (for identity injection, etc.) is on the roadmap for Kubernetes CSI before GA. By ephemeral I mean volumes who have their lifecycle directly tied to the consuming pod.

We've been looking for someone to try and implement such a volume and help us discover the gaps, so thanks for doing that.

To summarize, it sounds like you ran in to the following blockers:

1. P0 - CSI does not pass CO workload information to plugin (specifically the pod's name, namespace, and UID).
2. P1 - Ability to run node only volume plugins (no attach required) without a no-op external-attacher.
3  P2 - Provide documentation and/or libraries for writing ephemeral volumes.

We plan to close (at least) the P0 and P1 items before CSI will GA. We can use this bug as a tracking bug.

**Comment by smarterclayton** [comment]
I believe that those are all correct.  I would maybe phrase P1 is "having to create a PV and a PVC for every ephemeral volume may be too onerous" and/or "creating a PV to represent an ephemeral pod local volume is not correct".

**Comment by kfox1111** [comment]
P0 will solve one of my problems I'm trying to solve in:
https://github.com/kubernetes-csi/drivers/issues/85

I have the driver written, but want to use pod id (or namespace+podname) as the csi volumeid. so that each instance is pod specific.

Is anyone working on p0 currently?

There is also a p3: https://github.com/kubernetes/kubernetes/issues/64157 as you can use flexvolume in a pod directly.

**Comment by smarterclayton** [comment]
Your P3 is what I phrased as maybe my P1 - it should be possible to safely declare on a pod a node local volume (which needs no external state).  But we started with PV for security / safety reasons so we have to be careful to keep all of those in mind.

**Comment by kfox1111** [comment]
Hmm. ok. Thanks. :)

**Comment by jsafrane** [comment]
CSI volumes in-lined into pods are tracked in https://github.com/kubernetes/community/pull/2199. We definitely want to support them before GA.

**Comment by gnufied** [comment]
There is a ongoing discussion about whether CSI RPC calls should accept CO workload information -https://github.com/container-storage-interface/spec/issues/248 . I think the concern being raised there is -  if CSI plugins start accepting CO specific params then at least some of them will become CSI plugins that only work on particular CO.

**Comment by davidz627** [comment]
@jsafrane The proposal here https://github.com/kubernetes/community/pull/2199 does not cover the general in-line volume case, it only deals with "Migration" of current in-tree in-line volumes to call out to CSI instead.

The key point here is I am not proposing adding something new in the pod spec like:
```
volumes:
  - name: csi-storage
    csiVolume:
        sample-key: sample-value
```

rather just making 

```
volumes:
   - name: myvolume
     gcePersistentDisk:
         pdName: test-disk
         fsType: ext4
```
call out to the GCE PD CSI Driver instead.

If we want to support in-line volumes with CSI that is a separate proposal and I am not aware of any current designs for that.

**Comment by kfox1111** [comment]
Hmm. yeah. kubernetes/community#2199 is talking about transparently mapping existing api to csi drivers. not allowing them to be called like a csi driver. We need another proposal to cover csi in pod spec directly.

Though if the csi in pod spec issue was solved, then the mapping could be done by upgrading the specs to be csi inline instead of mapping them at runtime. Might be a simpler implementation in the long run.

**Comment by jsafrane** [comment]
I'll create a proposal for in-tree CSI volumes then.

**Comment by saad-ali** [comment]
> 2. Allow the contents of an image to be mounted into a pod (https://github.com/dims/docker-flexvol) - so the user defines a volume of type "container.image" that has the image name as a volume parameter ("quay.io/myapplication/static-assets:latest") that is mounted into a directory in my pod (at "/srv/static") so that I don't have to copy them out of an init container.

This can be handled today. The source image can be specified as `VolumeAttributes` and the plugin would fetch that image, unpack it, and put it at the mount path.

> 1. Allow a pod to be given a per node identity from a non-Kube identity provider - for example, a local daemon that provides Kerberos or Vault identity tokens as a directory in the container.
>
> This is a very opinionated custom secret volume that gets its source from someplace other than Kubernetes. The volume is explicitly associated with the node, the pod, and the service account. A cluster level controller might inject a flex volume of type "local.kerberos.identity" into the pod, and the local daemon would then provide a krb5.keytab file inside of a tmpfs mounted directory.

This maybe could be handled by kubelet, if we have a some flag indicating that the plugin needs "pod identity". Drawback is it would require kubelet to understand how to talk to third party secret store and where would we put the "requires pod identity switch". Will think on this more.

**Comment by kfox1111** [comment]
@saad-ali some of this is being discussed in https://github.com/kubernetes/community/pull/2273 too.

**Comment by kfox1111** [comment]
I've got a prototype implementation of a csi replacement of docker-flexvol here:
https://github.com/kfox1111/drivers/tree/volume-driver/pkg/image

It works as a csi driver, but need some plumbing on the k8s side to make it usable.

**Comment by jsafrane** [comment]
Regarding P0, passing pod namespace+name+uid to csi drivers: We can introduce set of attributes that kubelet passes in NodePublish.attributes when a driver requires so:

```go
volumeAttribs := <attributes from PV or in-line volume>

if driverRequiresPodInfo {
		volumeAttribs["kubernetes.io/podName"] = pod.Name
		volumeAttribs["kubernetes.io/podNamespace"] = pod.Namespace
		volumeAttribs["kubernetes.io/podUID"] = pod.UID
}

req := NodePublishVolumeRequest{
		VolumeId:           volID,
		TargetPath:         targetPath,
		VolumeAttributes:   volumeAttribs,
		...
}
```

These attributes would be Kubernetes specific and it's expected that drivers that require these parameters know something about Kubernetes, e.g. can GET the Pod based on the attributes.

Question is, how kubelet knows if a driver requires kubelet to pass pod information (i.e. where `driverRequiresPodInfo` comes from). There are several ways, none of them require CSI changes.

* For example, we could have a global CSIDriver API object that  would hold this as a bool flag (among other things; currently under discussion in kubernetes-csi community). This object would be created by some installation tool, e.g. helm chart or an operator, when installing a CSI driver.

* Or we can extend the CSI registration mechanism to pass this flag from CSI driver registrar to kubelet. Currently, it passes only path to driver CSI socket to kubelet. It could pass also a flag. CSI driver registrar is part of a DaemonSet that's installed by something (helm/operator/...) and it can create the DaemonSet so the registrar gets a command line argument (env. variable / config map / ...) with value of the flag.

In both cases it's up to the driver authors to document if this flag is required by the driver and/or provide the right helm charts.

**Comment by kfox1111** [comment]
@jsafrane was a decision made whether to use crd or registrar flag?

**Comment by jsafrane** [comment]
I wrote proposal how to pass pod name, namespace, uid and service account to a CSI driver: https://github.com/kubernetes/community/pull/2439, similarly to what Flex does.

I chose a new gRPC API betwen CSI driver registrar and kubelet. Right now it passes just a flag if the driver that's being registered wants pod information or not. In future it might pass something more.
